From 7e667664d28f90bf6916604a55ebad7e2d85305b Mon Sep 17 00:00:00 2001
From: Andrea Bolognani <abologna@redhat.com>
Date: Wed, 22 Mar 2017 13:44:13 +0100
Subject: [PATCH] qemu: Fix memory locking limit calculation

For guests that use <memoryBacking><locked>, our only option
is to remove the memory locking limit altogether.

Partially-resolves: https://bugzilla.redhat.com/1431793

Forwarded: no (backport)
Author: Jorge Niedbalski <jorge.niedbalski@canonical.com>
Original-Author: Andrea Bolognani <abologna@redhat.com>
Origin: https://libvirt.org/git/?p=libvirt.git;a=patch;h=7e667664d28f90bf6916604a55ebad7e2d85305b
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1708305
--- libvirt-1.3.1.orig/src/qemu/qemu_domain.c
+++ libvirt-1.3.1/src/qemu/qemu_domain.c
@@ -4095,6 +4095,14 @@ qemuDomainGetMemLockLimitBytes(virDomain
         goto done;
     }
 
+    /* If the guest wants its memory to be locked, we need to raise the memory
+     * locking limit so that the OS will not refuse allocation requests;
+     * however, there is no reliable way for us to figure out how much memory
+     * the QEMU process will allocate for its own use, so our only way out is
+     * to remove the limit altogether. Use with extreme care */
+    if (def->mem.locked)
+        return VIR_DOMAIN_MEMORY_PARAM_UNLIMITED;
+
     if (ARCH_IS_PPC64(def->os.arch)) {
         unsigned long long maxMemory;
         unsigned long long memory;
